beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -480,6 +480,24 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         self.metrics.attestation_processing_requests.inc();
         let timer = self.metrics.attestation_processing_times.start_timer();
 
+        // Retrieve the attestation's state from `store` if necessary.
+        let attestation_state = match attestation.data.beacon_block_root == self.canonical_head.read().beacon_block_root {
```

Comment:
This works, but we could loosen this `if` statement and do less reads from the database.

A `BeaconState` can be used to verify an `Attestation` if both of these conditions are true:

- The `BeaconState` and `Attestation` are "age appropriate". I.e., `epoch(attestation_slot)` is either equal or adjacent to `epoch(state.slot)`.
- The `BeaconState` block history contains the `target.root`. (You can use [`BestBlockRootsIterator`](https://github.com/sigp/lighthouse/blob/051355925299862ea2298c6f5d24fdf1dfc6fbd0/beacon_node/store/src/iter.rs#L165-L188) to do this quite easily).

If either of those two conditions are not met, then we use `target.root` to load the `BeaconBlock`, then use `block.state_root` to load a `BeaconState` that is necessarily a valid state to use. (Note: I think using `attestation.data.target.root` is better than `attestation.data.beacon_block_root`, as I had previously said).